README.md
<p align="center">
  <a href="" rel="noopener">
 <img width=200px height=200px src="cmrust1.jpg" alt="Project logo"></a>
</p>

<h3 align="center">Calculations_of_a_Circle</h3>

<div align="center">

[![Status](https://img.shields.io/badge/status-active-success.svg)]()

</div>

---

<p align="center">A simple demo of the use of the GTK4 package.
    <br> 
</p>

## 🧐 About <a name = "about"></a>


<p align="center">
  <a href="" rel="noopener">
 <img width=200px height=200px src="screen1.jpg" alt="Screen1"></a>
</p>
<p align="center">
  <a href="" rel="noopener">
 <img width=200px height=200px src="screen2.jpg" alt="Screen2"></a>
</p>

<p align="center">Screen shots of the app.
    <br> 
</p>

### Three input fields. 

1. The first takes the radius of the circle accepts once the submit button is pressed and displays answer when the calculate button is press.
2. The second takes the radious and calculates the circumference of a circle. with similar actions.
3. The final adds Terry to any text string.

The demo is built using the *container* method, where each block of the GUI is built within a *container* as shown in the code below.  
The variable _btn_ is the button, _with_label_ shows the adjacent text as the label for the button, _btn.connected_clicked_ will upon the _btn_ being clicked grab the text in the screen as a **String**, *text.trim().parse::<f32>().unwrap()* causes the **String** of text to be a type f32.


```rust
let from_entry = gtk::Entry::builder()
    .placeholder_text("Radius")
    .build();
text_container.append(&from_entry);
let btn = gtk::Button::with_label("Enter");
btn.connect_clicked(clone!(@weak clipboard, @weak from_entry => move |_btn| {
    let text = from_entry.text();
    let radius: f32 = text.trim().parse::<f32>().unwrap();
    let circum_circle: f32 = radius * 2. * 3.1415;
    let circumanswer = circum_circle.to_string();
    clipboard.set_text(&circumanswer)
   }));
```

## 🏁 Purpose and Getting Going <a name = "getting_started"></a>

The main purpose of this demo is to provide a code template for creating GUIs for RUST code projects moving forward.  This is really part one of two.  In this first demo/template the layout is constructed in a strict Box Builder container style. Using this paradaigm each element is added in a container stacked progressively downward in the window. The other builder approach not included here is a grid builder pattern.  These builder patterns rely on constructed pre-defined widgets which can be selected from [here](https://docs.gtk.org/gtk4/visual_index.html)

Alternatively a non-builder pattern can be used in gtk which they call a Composite Template.  Where GTK allows for a custom layout to describe the user interface with a markup language - they use XML. see the documentation at (https://gtk-rs.org/gtk4-rs/stable/latest/book/composite_templates.html) 


### Build
  
Build using the "cargo run" command.

### Docs
  
Using "cargo doc" generates the documentation in the .../src/target/docs directory.

### Debug
  
Automatically sets up debugging on the MacOS using VScode and LLDB Code extensions.

#### Prerequisites

Need Vscode with extensions: 
- rust-analyzer
- code lldb

## ⛏️ Built Using <a name = "built_using"></a>

- [vscode](https://www.vscode.com)
- [rust](https://www.rustlang.org)
- [gtk-rs](https://gtk-rs.org)

## ✍️ Author <a name = "author"></a>

- [@iotandwearablesguy](https://github.com/iotandwearablesguy) - Idea & Initial work

## 🎉 Acknowledgements <a name = "acknowledgement"></a>

- [Thank You GTK-RS]([https://simplemaps.com/data/us-cities](https://gtk-rs.org))



 ## LICENSE

_O'Shea-Dishongh No Harm License© 2022 Terrance O'Shea & Katherine Dishongh_

<http://github.com/terryo-iotandwearablesguy>

Subject to the following conditions:

 1. That the below copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

 2. That any use of the Software under this license never be used for harmful purposes, permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction,including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to
 do so subject to the aforementioned restrictions.

"Harmful purposes" as used herein means to not harm destroy or otherwise restrict humans, animals, or the environment. "Harmful purposes" includes, inter alia, the creation of weaponry, the guidance of weapons, limitation of food and services, damaging, contaminating or otherwise hazardous to the environmnet in any form, and/or restricting the freedom or movement of people.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE






CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Released] - 2022-11-12

Initial release with the proper documentation and readme.md files

### Added


Three input fields. 

1. The first takes the radius of the circle accepts once the submit button is pressed and displays answer when the calculate button is press.
2. The second takes the radious and calculates the circumference of a circle. with similar actions.
3. The final adds Terry to any text string.

### TODOs Future Releases

1. Fix vscode templates
2. Set up actions and clean up interfaces CSS.
3. Clean up variable names.


### Changed

- initial release

### Deprecated

- none.

### Removed

- none.

### Fixed

- none.

### Security

- Vulnerabilities - unknown.


src/main.rs
use gdk::Display;
use glib::clone;
#[allow(unused_imports)]
use gtk::{
 CssProvider, StyleContext, gio, glib, Application, ApplicationWindow, Button, Label, prelude::*, gdk,
};
fn main() {
    
    let application = gtk::Application::new(
        Some("com.github.gtk-rs.examples.clipboard"),
        Default::default(),
    );
       
    application.connect_startup(|_| load_css());
    application.connect_activate(build_ui);
    application.run();
}
/// This fn *is* [`load_css`]!
fn load_css() {
    // Load the CSS file and add it to the provider
    let provider = CssProvider::new();
    provider.load_from_data(include_bytes!("style.css"));

    // Add the provider to the default screen
    StyleContext::add_provider_for_display(
        &Display::default().expect("Could not connect to a display."),
        &provider,
        gtk::STYLE_PROVIDER_PRIORITY_APPLICATION,
    );
}
fn build_ui(application: &gtk::Application) {
    let window = gtk::ApplicationWindow::builder()
        .application(application)
        .title("Calculations of a Circle")
        .default_width(400)
        .default_height(300)
        .build();

    let display = gdk::Display::default().unwrap();
    let clipboard = display.clipboard();

    let container = gtk::Box::builder()
        .orientation(gtk::Orientation::Vertical)
        .margin_top(24)
        .margin_bottom(24)
        .margin_start(24)
        .margin_end(24)
        .halign(gtk::Align::Center)
        .valign(gtk::Align::Center)
        .spacing(24)
        .build();

    // The text copy/paste part
    //let outputcir =format!("Distance between {} and {} on the surface of Earth is  {:.1} miles.", city1, city2, miles_distance );
    let title = gtk::Label::builder()
        //.label("Radius                                                          Circumference")
        .label("Circumference of a Circle")
        .halign(gtk::Align::Start)
        .build();
    title.add_css_class("title-2");
    container.append(&title);

    let text_container = gtk::Box::builder()
        .halign(gtk::Align::Center)
        .orientation(gtk::Orientation::Horizontal)
        .spacing(24)
        .build();

    let from_entry = gtk::Entry::builder()
        .placeholder_text("Radius")
        .build();
    text_container.append(&from_entry);

    let copy_btn = gtk::Button::with_label("Enter");
    copy_btn.connect_clicked(clone!(@weak clipboard, @weak from_entry => move |_btn| {
        let text = from_entry.text();
        let radius: f32 = text.trim().parse::<f32>().unwrap();
        let circum_circle: f32 = radius * 2. * 3.1415;
        let circumanswer = circum_circle.to_string();
        clipboard.set_text(&circumanswer);
    }));
      copy_btn.set_widget_name("button_enter");
    text_container.append(&copy_btn);

    let into_entry = gtk::Entry::new();
    text_container.append(&into_entry);

    let paste_btn = gtk::Button::with_label("Result");
    paste_btn.connect_clicked(clone!(@weak clipboard, @weak into_entry => move |_btn| {
        clipboard.read_text_async(gio::Cancellable::NONE, clone!(@weak into_entry => move|res| {
            if let Ok(Some(circumanswer)) = res {
                into_entry.set_text(&circumanswer);
            }
        }));
    }));
      paste_btn.set_widget_name("button_results");
    text_container.append(&paste_btn);
    container.append(&text_container);
//
//
let title3 = gtk::Label::builder()
     //   .label("Radius                                                           Area")
    .label("Area of a Circle")
        .halign(gtk::Align::Start)
        .build();
    title3.add_css_class("title-2");
    container.append(&title3);

    let text_container3 = gtk::Box::builder()
        .halign(gtk::Align::Center)
        .orientation(gtk::Orientation::Horizontal)
        .spacing(24)
        .build();

    let from_entry3 = gtk::Entry::builder()
        .placeholder_text("Enter the radius of Circle")
        .build();
    text_container3.append(&from_entry3);

    let copy_btn3 = gtk::Button::with_label("Enter");
    copy_btn3.connect_clicked(clone!(@weak clipboard, @weak from_entry => move |_btn| {
        let text3 = from_entry3.text();
        let radius: f32 = text3.trim().parse::<f32>().unwrap();
        let areacircle: f32 = radius * radius * 3.1415;
        let area_answer = areacircle.to_string();
        clipboard.set_text(&area_answer);
    }));
      copy_btn3.set_widget_name("button_enter");
    text_container3.append(&copy_btn3);

    let into_entry = gtk::Entry::new();
    text_container3.append(&into_entry);

    let paste_btn3 = gtk::Button::with_label("Result");
    paste_btn3.connect_clicked(clone!(@weak clipboard, @weak into_entry => move |_btn| {
        clipboard.read_text_async(gio::Cancellable::NONE, clone!(@weak into_entry => move|res| {
            if let Ok(Some(area_answer)) = res {
                into_entry.set_text(&area_answer);
            }
        }));
    }));
      paste_btn3.set_widget_name("button_results");
    text_container3.append(&paste_btn3);
    container.append(&text_container3);
//
//
    let title2 = gtk::Label::builder()
       // .label("Text Entry                                                      Added")
        .label("Text Adder Example").halign(gtk::Align::Start)
        .build();
    title2.add_css_class("title-2");
    container.append(&title2);

    let text_container2 = gtk::Box::builder()
        .halign(gtk::Align::Center)
        .orientation(gtk::Orientation::Horizontal)
        .spacing(24)
        .build();

    let from_entry2 = gtk::Entry::builder()
        .placeholder_text("Text Adder")
        .build();
    text_container2.append(&from_entry2);

    let copy_btn2 = gtk::Button::with_label("Enter");
    copy_btn2.connect_clicked(clone!(@weak clipboard, @weak from_entry => move |_btn| {
        let text2 = from_entry2.text();
        let terry_text = format!{"Terry "} + &text2;
        clipboard.set_text(&terry_text);
    }));
      copy_btn2.set_widget_name("button_enter");
    text_container2.append(&copy_btn2);

    let into_entry2 = gtk::Entry::new();
    text_container2.append(&into_entry2);

    let paste_btn2 = gtk::Button::with_label("Result");
    paste_btn2.connect_clicked(clone!(@weak clipboard, @weak into_entry => move |_btn| {
        clipboard.read_text_async(gio::Cancellable::NONE, clone!(@weak into_entry2 => move|res| {
            if let Ok(Some(terry_text)) = res {
                into_entry2.set_text(&terry_text);
            }
        }));
    }));
        paste_btn2.set_widget_name("button_results");
    text_container2.append(&paste_btn2);
    container.append(&text_container2);
    //
    //
        let titleq = gtk::Label::builder()
            .label("Quit")
        .halign(gtk::Align::Center)
        .build();
    titleq.add_css_class("title-2");
    container.append(&titleq);

    let text_containerq = gtk::Box::builder()
        .halign(gtk::Align::Center)
        .orientation(gtk::Orientation::Horizontal)
        .spacing(24)
        .build();

    let button_4 = gtk::Button::with_label("Quit");
    button_4.connect_clicked(clone!(@weak window => move |_| window.destroy()));
    button_4.add_css_class("destructive-action");

    text_containerq.append(&button_4);
     container.append(&text_containerq);
    window.set_child(Some(&container));
    window.show();
}


src/lib.rs
//!    **Calculations of a Circle Application_ App** 
//!             _© 2022 Terrance O'Shea_
//!             _terry.oshea@me.com_
//! ## Intro:
//! 
//! Example of GTK4 Package Using Calculation of a Circle and Text Augmenting."
//!
//! ## Acknowledgements:
//! 
//! * The Gui's are made with gtk4
//! 
//! * The application reads from 
//! 
//! ```
//! let from_entry = gtk::Entry::builder()
//!     .placeholder_text("Radius")
//!     .build();
//! text_container.append(&from_entry);
//! let btn = gtk::Button::with_label("Enter");
//! btn.connect_clicked(clone!(@weak clipboard, @weak from_entry => move |_btn| {
//!     let text = from_entry.text();
//!     let radius: f32 = text.trim().parse::<f32>().unwrap();
//!     let circum_circle: f32 = radius * 2. * 3.1415;
//!     let circumanswer = circum_circle.to_string();
//!     clipboard.set_text(&circumanswer)
//!   }));
//! ```
//! 
//!  ## LICENSE
//! 
//! _O'Shea-Dishongh No Harm License© 2022 Terrance O'Shea & Katherine Dishongh_
//! 
//! <http://github.com/terryo-iotandwearablesguy>
//! 
//! Subject to the following conditions:
//! 
//!  1. That the below copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
//! 
//!  2. That any use of the Software under this license never be used for harmful purposes, permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction,including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to
//!  do so subject to the aforementioned restrictions.
//! 
//! "Harmful purposes" as used herein means to not harm destroy or otherwise restrict humans, animals, or the environment. "Harmful purposes" includes, inter alia, the creation of weaponry, the guidance of weapons, limitation of food and services, damaging, contaminating or otherwise hazardous to the environmnet in any form, and/or restricting the freedom or movement of people.
//! 
//! THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE
//! 
//! 
//! 
//!
pub fn load_css() {
//! loads from style.css
}
#[allow(dead_code)]
pub fn build_ui(application: &gtk::Application) {
#[allow(unused_variables)]   
    let window = gtk::ApplicationWindow::builder()
        .application(application)
        .title("Calculations of a Circle")
        .default_width(400)
        .default_height(300)
        .build();
}
/*fn set_panic_hook() {
    #[cfg(target_arch = "wasm32")]
    console_error_panic_hook::set_once();
}

pub fn answer() -> u32 {
    42
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deep_thought_test() {
        assert_eq!(answer(), 42);
    }
}
*/
